### ISSUE TITLE
Add scope type "pattern"

### ISSUE THREAD

**Issue by wenkokke** [cursorless-dev/cursorless#2177]
I propose we add a scope type for the patterns in pattern matches, which are very important in functional languages like Haskell, Scala, and Rust, and increasingly important in languages such as JavaScript (see [tc39](https://github.com/tc39/proposal-pattern-matching)) and Python (see [pep 636](https://peps.python.org/pep-0636/)).

```typescript
// TypeScript
function fst<A, B>(tup: [A, B]) {
  const { x, y } = tup;
  //    ^^^^^^^^
  return x;
}
// Using tc39
function fst<A, B>(tup: [A, B]) {
  match (tup) {
    when ({ x, y }): return x;
    //    ^^^^^^^^
  }
}
```

```python
# Python
def fst(tup):
  match tup:
    case (x, y): return x
#        ^^^^^^
```

```scala
// Scala
def fst[A, B](tup:  (A, B)): A =
  tup match {
    case (x, y) => x
    //   ^^^^^^
  }
```

```haskell
-- Haskell
fst :: (a, b) -> a
fst (x, y) = x
--  ^^^^^^
```

```rust
// Rust
fn fst(tup: (A, B)) {
  match tup {
    (x, y) => x
//  ^^^^^^
  }
}
```

Patterns are different from conditions. The easiest way to show this is by implementing the if then else construct using patterns.

```python
def ifThenElse(cond, ifTrue, ifFalse):
  match cond:
#       ^^^^  <-- "condition"
    case True:
#        ^^^^  <-- "pattern"
      return ifTrue()
    case False:
#        ^^^^^  <-- "pattern"
      return ifFalse()
```

Patterns are not adequately covered by branches, since "branch" selects the entire branch, which may match on multiple patterns.

```haskell
and ::  Bool -> Bool -> Bool
and True True = True
--  ^^^^^^^^^^^^^^^^  <-- branch
--  ^^^^  <-- pattern
--       ^^^^  <-- pattern
and _    _    = False
--  ^^^^^^^^^^^^^^^^^ <-- branch
--  ^  <-- pattern
--       ^  <-- pattern
```

**Comment by wenkokke** [comment]
@pokey wrote [here](https://github.com/cursorless-dev/cursorless/issues/579#issuecomment-1161627378) that the patterns should be addressed by "name", "key", or "condition".

I believe I've shown above that "condition" should refer to the "subject".

Furthermore, the complexity of patterns makes it rather alient to refer to them by "name" or "key". That suggestion may have been due to the overly simplistic nature of the examples. For instance, see the following Haskell program, where I've annotated the top-most patterns. At very least, I hope that shows that "name" is inadequate?

```haskell
someFunction x (y1 : y2 : ys) (a, b, (c, [d])) = rhs
--           ^                                  <-- pattern
--             ^^^^^^^^^^^^^^                   <-- pattern
--                            ^^^^^^^^^^^^^^^^  <-- pattern
```

Note: The program looks contrived, due to the fact I've copied it from a parser test. However, such patterns, albeit with better names, are extremely common in Haskell.

**Comment by pokey** [comment]
my thinking would be


```haskell
someFunction x (y1 : y2 : ys) (a, b, (c, [d])) = rhs
--           ^ ^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ <-- "arg"
--              ^^   ^^   ^^   ^  ^  ^^^^^^^^  <-\ "item"
--                                    ^  ^^^   <-|
--                                        ^    <-/
```

**Comment by wenkokke** [comment]
I'm not sure if item makes sense in the general case?

```haskell
either                  :: (a -> c) -> (b -> c) -> Either a b -> c
either f _ (Left x)     =  f x
either _ g (Right y)    =  g y
```

Are the `x` and `y` on the left hand side still item?

**Comment by wenkokke** [comment]
@pokey What do you think should be used in the JavaScript, Python and Rust examples?

**Comment by wenkokke** [comment]
These aren't `value` because they're binding occurrences rather than values, so the use cases, domain and iteration scopes are very different.

These aren't `key` or `item` because they're not always in a list or map like context, and furthermore because they're binders rather than items in a list.

**Comment by pokey** [comment]
> I'm not sure if item makes sense in the general case?
> 
> ```haskell
> either                  :: (a -> c) -> (b -> c) -> Either a b -> c
> either f _ (Left x)     =  f x
> either _ g (Right y)    =  g y
> ```
> 
> Are the `x` and `y` on the left hand side still item?

no I think I'd call those `"arg"`?

**Comment by pokey** [comment]
we tend to use the same scope name whether it's constructing or deconstructing. makes it easier for the ol' lizard brain that just sees shapes

**Comment by pokey** [comment]
> @pokey What do you think should be used in the JavaScript, Python and Rust examples?

I believe they're as follows, but easiest to paste them into a file, pull up the "bar cursorless" sidebar and click through the scopes to see what hits

```typescript
// TypeScript
function fst<A, B>(tup: [A, B]) {
  const { x, y } = tup;
  //    ^^^^^^^^  <~~ name
  return x;
}
// Using tc39
function fst<A, B>(tup: [A, B]) {
  match (tup) {
    when ({ x, y }): return x;
    //    ^^^^^^^^ <~~ condition
  }
}
```

```python
# Python
def fst(tup):
  match tup:
    case (x, y): return x
#        ^^^^^^ <~~ condition
```

**Comment by wenkokke** [comment]
> > I'm not sure if item makes sense in the general case?
> 
> > 
> 
> > ```haskell
> 
> > either                  :: (a -> c) -> (b -> c) -> Either a b -> c
> 
> > either f _ (Left x)     =  f x
> 
> > either _ g (Right y)    =  g y
> 
> > ```
> 
> > 
> 
> > Are the `x` and `y` on the left hand side still item?
> 
> 
> 
> no I think I'd call those `"arg"`?

So it's nested args all the way down? Because `(Left x)` is also "arg".

**Comment by pokey** [comment]
> So it's nested args all the way down? Because `(Left x)` is also "arg".

In this case, yes, but note that in https://github.com/cursorless-dev/cursorless/issues/2177#issuecomment-1898749525 there's no nesting

**Comment by wenkokke** [comment]
True, but item is inappropriate there, so those cases would also be "arg".

**Comment by pokey** [comment]
> item is inappropriate there

can you elaborate?

**Comment by wenkokke** [comment]
> > item is inappropriate there
> 
> can you elaborate?

If

```haskell
foo x = undefined
--  ^ arg
```

and

```haskell
foo (Left x) = undefined
--        ^ arg
```

then

```haskell
foo (x : xs) = undefined
--   ^ arg
```

and 

```haskell
foo [x] = undefined
--   ^ arg
```

because they're all the same kind of thing.

**Comment by wenkokke** [comment]
One major use case for this might be fixed by separating `@argument.formal` and `@argument.actual`.

**Comment by pokey** [comment]
interesting. maybe let's discuss this one in meet-up?
